<launch>
    
    <arg name="use_dron" default="true"/>
    <arg name="mission_enabled" default="false"/> <!-- To use mission_interface node-->
    <arg name="hmi" default="false"/> 
    
    <arg name="lidar_topic" default="os1_cloud_node/points"/>
    <arg name="use_octomap" default="true"/>
    <arg name="verbose_optimizer" default="false"/>
    <arg name="debug_msgs" default="true"/>
    <arg name="show_config_msgs" default="true"/>

        <!-- Mission Planner -->
        <node pkg="theta_star_2d" type="action_goal_interface_node" name="mission_interface">
            <remap from="/Execute_Mission" to="Execute_Mission"/>
            <remap from="/Make_Plan" to="Make_Plan"/>
            <param name="hmi_mode" value="$(arg hmi)"/>
            <param name="mission_enabled" value="$(arg mission_enabled)"/>
            <param name="waiting_time" value="5.0"/>
        </node>

        <!-- Theta* Planner -->
        <node pkg="theta_star_2d" type="global_planner_node" name="global_planner_node" output="screen"> # launch-prefix="gdb -ex run --args">
            <rosparam command="load" file="$(find marsupial_g2o)/cfg/3d_global_params.yaml"/>
            <remap from="/Make_Plan" to="Make_Plan"/>
            <remap from="/Execute_Plan" to="Execute_Plan"/>
            <remap from="/points" to="mcl3d_node/map_point_cloud" if="$(eval arg('use_dron') and not arg('use_octomap'))" />
            <param name="mode3d" value="$(arg use_dron)"/>
            <param name="use_octomap" value="$(arg use_octomap)"/>
            <param name="debug" value="$(arg debug_msgs)"/>
            <param name="robot_base_frame" value="uav_base_link"/>
            <param name="timeout" value="30"/>
        </node>

        <node pkg="marsupial_g2o" type="optimizer_local_planner_node" name="optimizer_local_planner_node" output="screen"> # launch-prefix="gdb -ex run --args">
            <rosparam command="load" file="$(find marsupial_g2o)/cfg/3d_local_params.yaml"/>
            <remap from="/Execute_Plan" to="Execute_Plan"/>
            <remap from="/Navigation" to="Navigation"/>
            <remap from="/points" to="$(arg lidar_topic)" if="$(eval not arg('use_octomap') and arg('use_dron'))" />
            <remap from="/octomap_binary_local" to="/octomap_full_local" if="$(eval arg('use_dron') and arg('use_octomap'))"/>
            <param name="debug" value="$(arg debug_msgs)"/>
            <param name="show_config" value="$(arg show_config_msgs)"/>
            <param name="mode3d" value="$(arg use_dron)"/>
            <param name="uav_base_frame" value="uav_base_link"/>
            <param name="ugv_base_frame" value="ugv_base_link"/>
            <param name="traj_dest_frame" value="uav_base_link"/>
            <param name="timeout" value="0.4"/>
            <param name="final_search_around" value="1.4"/>
            
            <param name="verbose_optimizer" value="$(arg verbose_optimizer)"/>
            
            <param name="initial_pos_ugv_x"  value="-3.0"/>   <!-- distance-vertex 5.44--> 
            <param name="initial_pos_ugv_y"  value="6.0"/>   <!-- distance-xyz 5.16-->
            <param name="initial_pos_ugv_z"  value="0.2"/>   <!-- equi-distance 4.78-->

            <param name="w_alpha"   value="0.1"/>   <!-- distance-vertex 5.44--> 
            <param name="w_gamma"   value="0.1"/>   <!-- distance-xyz 5.16-->
            <param name="w_epsilon" value="0.8"/>   <!-- equi-distance 4.78-->
            <param name="w_beta"    value="0.6"/>   <!-- obstacles 27.71-->
            <param name="w_iota"    value="0.4"/>   <!-- obstacles through -->
            <param name="w_delta"   value="0.9"/>   <!-- kinematics 5.52-->
            <param name="w_zeta"    value="0.05"/>   <!-- time -->
            <param name="w_eta"     value="0.3"/>   <!-- velocity -->
            <param name="w_theta"   value="1.0"/>   <!-- acceleration -->
            <param name="n_iter_opt"    value="500"/>   <!-- number max. of iterations per optimization --> 
            <param name="n_time_optimize"    value="1"/>   <!-- number Loop-For to execute optimizer --> 
            <param name="td_"    value="1.0"/>   <!-- Time Sleep between optimization in Loop-For --> 

            <param name="bound"         value="2.0"/> <!-- Bound distance to obstacles -->
            <param name="velocity"      value="2.0"/>    
            <param name="acceleration"  value="0.0"/>    
        </node>

</launch>
